javapoet/src/main/java/com/palantir/javapoet/TypeSpec.java

```diff
@@ -255,96 +272,114 @@ public Builder toBuilder() {
+            codeWriter.emit("$L", enumName);
+            if (!anonymousTypeArguments.formatParts().isEmpty()) {
+                codeWriter.emit("(");
                 codeWriter.emit(anonymousTypeArguments);
-                codeWriter.emit(") {\n");
+                codeWriter.emit(")");
+            }
+            if (fieldSpecs.isEmpty() && methodSpecs.isEmpty() && typeSpecs.isEmpty()) {
+                return false; // Avoid unnecessary braces "{}".
+            }
+            codeWriter.emit(" {\n");
+        } else if (anonymousTypeArguments != null) {
+            TypeName supertype = !superinterfaces.isEmpty() ? superinterfaces.get(0) : superclass;
+            codeWriter.emit("new $T(", supertype);
+            codeWriter.emit(anonymousTypeArguments);
+            codeWriter.emit(") {\n");
+        } else {
+            // Push an empty type (specifically without nested types) for type-resolution.
+            codeWriter.pushType(new TypeSpec(this));
+
+            codeWriter.emitJavadoc(
+                    kind == Kind.RECORD ? MethodSpec.makeJavadocWithParameters(javadoc, recordComponents) : javadoc);
+            codeWriter.emitAnnotations(annotations, false);
+            codeWriter.emitModifiers(modifiers, Util.union(implicitModifiers, kind.asMemberModifiers));
+            if (kind == Kind.ANNOTATION) {
+                codeWriter.emit("$L $L", "@interface", name);
             } else {
-                // Push an empty type (specifically without nested types) for type-resolution.
-                codeWriter.pushType(new TypeSpec(this));
-
-                codeWriter.emitJavadoc(javadoc);
-                codeWriter.emitAnnotations(annotations, false);
-                codeWriter.emitModifiers(modifiers, Util.union(implicitModifiers, kind.asMemberModifiers));
-                if (kind == Kind.ANNOTATION) {
-                    codeWriter.emit("$L $L", "@interface", name);
-                } else {
-                    codeWriter.emit("$L $L", kind.toString(), name);
-                }
-                codeWriter.emitTypeVariables(typeVariables);
+                codeWriter.emit("$L $L", kind.name().toLowerCase(Locale.US), name);
+            }
+            codeWriter.emitTypeVariables(typeVariables);
 
-                List<TypeName> extendsTypes;
-                List<TypeName> implementsTypes;
-                if (kind == Kind.INTERFACE) {
-                    extendsTypes = superinterfaces;
-                    implementsTypes = Collections.emptyList();
-                } else {
-                    extendsTypes = superclass.equals(ClassName.OBJECT)
-                            ? Collections.emptyList()
-                            : Collections.singletonList(superclass);
-                    implementsTypes = superinterfaces;
-                }
+            // Record components.
+            if (kind == Kind.RECORD) {
+                MethodSpec.emitParameters(codeWriter, recordComponents, varargs);
```

Comment:
This isn't great. We're breaking the abstraction and reaching into something that should be internal to `MethodSpec`. I would much prefer we just duplicate this code here in `TypeSpec`.

Reply:
I think this probably makes the most sense as a method on `CodeWriter`, similar to the `emitTypeVariables` method used above.